cli/cli_app.go

```diff
@@ -384,7 +384,31 @@ func RunTerragrunt(terragruntOptions *options.TerragruntOptions) error {
 		}
 	}
 
-	return runTerragruntWithConfig(terragruntOptions, terragruntConfig, false)
+	isInitCommand := util.FirstArg(terragruntOptions.TerraformCliArgs) == CMD_INIT
+
+	err, output := runTerragruntWithConfig(terragruntOptions, terragruntConfig, false)
+
+	// If running Terraform resulted in an error that matches one of the init errors,
+	// then run the required interim init, and then re-run previous command as-is,
+	// otherwise return error without making any attempt to recover from it or
+	// retry anything.
+	if err != nil && isErrorRequiringInit(output, err, terragruntOptions) && !isInitCommand {
+		terragruntOptions.Logger.Println("Detected an error that requires Terraform to be re-initialized; running 'terraform init' now...")
+
+		// Run the "terraform init" command.
+		if err = runTerraformInit(terragruntOptions, terragruntConfig, nil); err != nil {
+			return err
+		}
+
+		// If the init succeeded, re-run previous command again.
+		if err, _ = runTerragruntWithConfig(terragruntOptions, terragruntConfig, false); err != nil {
+			return err
+		}
+	} else if err != nil {
+		return err
+	}
+
+	return nil
```

Comment:
Done. Thank you!

Reply:
Actually. We can't do that. Perhaps different refactor to make this look a little better.

Since if the initial `runTerragruntWithConfig()` returns an error, then it will be set for the remained of the runtime, thus in the end Terragrunt would report an issue even though we recovered (by running `terraform init` successfully):

```bash
[terragrunt] 2020/09/11 19:53:59 Hit multiple errors:
exit status 1
```

Reply:
Re-factored.